File: tools/sdks/compose/src/main/kotlin/io/thunder/compose/components/presentation/auth/SignIn.kt
package io.thunder.compose.components.presentation.auth

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.BasicText
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import io.thunder.android.*
import io.thunder.compose.LocalThunder
import io.thunder.compose.ThunderState
import io.thunder.compose.components.actions.BaseThunderSignInButton
import kotlinx.coroutines.launch

/** State passed to the [BaseThunderSignIn] builder slot. */
@Stable
class ThunderSignInState {
    var inputs by mutableStateOf<List<FlowInput>>(emptyList())
        internal set
    var actions by mutableStateOf<List<FlowAction>>(emptyList())
        internal set
    var isLoading by mutableStateOf(false)
        internal set
    var error by mutableStateOf<String?>(null)
        internal set

    internal var flowId: String? = null
    internal var onSubmit: (String) -> Unit = {}
    private val fieldValues = mutableStateMapOf<String, String>()

    fun fieldValue(name: String): String = fieldValues[name] ?: ""
    fun setField(name: String, value: String) { fieldValues[name] = value }
    fun fields(): Map<String, String> = fieldValues.toMap()
    fun submit(actionId: String) = onSubmit(actionId)

    internal fun update(response: EmbeddedFlowResponse) {
        flowId = response.flowId
        inputs = response.data?.inputs ?: emptyList()
        actions = response.data?.actions ?: emptyList()
    }
}

/** Full app-native sign-in form (spec §8.4 Presentation). */
@Composable
fun ThunderSignIn(
    applicationId: String,
    modifier: Modifier = Modifier,
    onComplete: (() -> Unit)? = null,
    onError: ((String) -> Unit)? = null,
) {
    val thunderState = LocalThunder.current
    val i18n = thunderState.i18n
    BaseThunderSignIn(applicationId = applicationId, modifier = modifier, onComplete = onComplete, onError = onError) { signInState ->
        Column(modifier = Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
            BasicText(i18n.resolve("signIn.title"))
            signInState.error?.let { BasicText(it) }
            signInState.inputs.forEach { input ->
                BasicTextField(
                    value = signInState.fieldValue(input.name),
                    onValueChange = { signInState.setField(input.name, it) },
                    modifier = Modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp)
                        .semantics { contentDescription = input.name },
                )
            }
            signInState.actions.forEach { action ->
                BaseThunderSignInButton(
                    label = action.label ?: i18n.resolve("signIn.submit"),
                    isLoading = signInState.isLoading,
                ) { signInState.submit(action.id) }
            }
            if (signInState.isLoading) BasicText(i18n.resolve("signIn.loading"))
        }
    }
}

/** Unstyled base variant (spec §8.3). */
@Composable
fun BaseThunderSignIn(
    applicationId: String,
    modifier: Modifier = Modifier,
    onComplete: (() -> Unit)? = null,
    onError: ((String) -> Unit)? = null,
    content: @Composable (ThunderSignInState) -> Unit,
) {
    val thunderState = LocalThunder.current
    val scope = rememberCoroutineScope()
    val signInState = remember { ThunderSignInState() }

    signInState.onSubmit = { actionId ->
        scope.launch {
            signInState.isLoading = true
            signInState.error = null
            try {
                val payload = EmbeddedSignInPayload(
                    flowId = signInState.flowId, actionId = actionId, inputs = signInState.fields()
                )
                val request = EmbeddedFlowRequestConfig(applicationId, FlowType.AUTHENTICATION)
                val response = thunderState.client.signIn(payload = payload, request = request)
                handleSignInResponse(response, signInState, thunderState, onComplete, onError)
            } catch (e: Exception) {
                signInState.error = e.message
                onError?.invoke(e.message ?: "Sign-in failed")
            } finally {
                signInState.isLoading = false
            }
        }
    }

    LaunchedEffect(applicationId) {
        signInState.isLoading = true
        try {
            val request = EmbeddedFlowRequestConfig(applicationId, FlowType.AUTHENTICATION)
            val payload = EmbeddedSignInPayload(actionId = "__initiate__")
            val response = thunderState.client.signIn(payload = payload, request = request)
            handleSignInResponse(response, signInState, thunderState, onComplete, onError)
        } catch (e: Exception) {
            signInState.error = e.message
            onError?.invoke(e.message ?: "Sign-in failed")
        } finally {
            signInState.isLoading = false
        }
    }

    Box(modifier = modifier) { content(signInState) }
}

private suspend fun handleSignInResponse(
    response: EmbeddedFlowResponse,
    signInState: ThunderSignInState,
    thunderState: ThunderState,
    onComplete: (() -> Unit)?,
    onError: ((String) -> Unit)?,
) {
    when (response.flowStatus) {
        FlowStatus.COMPLETE -> { thunderState.refresh(); onComplete?.invoke() }
        FlowStatus.PROMPT_ONLY -> signInState.update(response)
        FlowStatus.ERROR -> {
            val msg = response.failureReason ?: "Sign-in failed"
            signInState.error = msg
            onError?.invoke(msg)
        }
    }
}

File: tools/sdks/thunderid-swiftui/Sources/ThunderIDSwiftUI/components/presentation/auth/SignIn.swift
import SwiftUI
import ThunderID

/// Full app-native sign-in form. Drives the Flow Execution API loop (spec §8.4 Presentation).
public struct SignIn: View {
    @EnvironmentObject private var state: ThunderState
    @EnvironmentObject private var i18n: ThunderI18n
    public let applicationId: String
    public let onComplete: (() -> Void)?
    public let onError: ((String) -> Void)?

    public init(
        applicationId: String,
        onComplete: (() -> Void)? = nil,
        onError: ((String) -> Void)? = nil
    ) {
        self.applicationId = applicationId
        self.onComplete = onComplete
        self.onError = onError
    }

    public var body: some View {
        BaseSignIn(applicationId: applicationId, onComplete: onComplete, onError: onError) { signInState in
            VStack(alignment: .leading, spacing: 12) {
                Text(i18n.resolve("signIn.title"))
                    .accessibilityAddTraits(.isHeader)
                if let error = signInState.error {
                    Text(error).foregroundColor(.red)
                }
                ForEach(signInState.inputs, id: \.name) { input in
                    if input.type == "password" {
                        SecureField(input.name, text: signInState.binding(for: input.name))
                            .accessibilityLabel(input.name)
                            .frame(minHeight: 44)
                    } else {
                        TextField(input.name, text: signInState.binding(for: input.name))
                            .accessibilityLabel(input.name)
                            .frame(minHeight: 44)
                    }
                }
                ForEach(signInState.actions, id: \.id) { action in
                    Button(action.label ?? i18n.resolve("signIn.submit")) {
                        signInState.submit(actionId: action.id)
                    }
                    .disabled(signInState.isLoading)
                    .accessibilityLabel(action.label ?? i18n.resolve("signIn.submit"))
                    .frame(minWidth: 44, minHeight: 44)
                }
                if signInState.isLoading {
                    Text(i18n.resolve("signIn.loading"))
                }
            }
            .padding()
        }
    }
}

/// State container passed to the BaseSignIn builder.
@MainActor
public final class SignInState: ObservableObject {
    @Published public private(set) var inputs: [FlowInput] = []
    @Published public private(set) var actions: [FlowAction] = []
    @Published public private(set) var isLoading: Bool = false
    @Published public private(set) var error: String?

    private var fieldValues: [String: String] = [:]
    private var flowId: String?
    let submitAction: (String, [String: String], String?) -> Void

    init(submit: @escaping (String, [String: String], String?) -> Void) {
        self.submitAction = submit
    }

    public func binding(for name: String) -> Binding<String> {
        Binding(
            get: { self.fieldValues[name] ?? "" },
            set: { self.fieldValues[name] = $0 }
        )
    }

    public func submit(actionId: String) {
        submitAction(actionId, fieldValues, flowId)
    }

    func update(from response: EmbeddedFlowResponse) {
        flowId = response.flowId
        inputs = response.data?.inputs ?? []
        actions = response.data?.actions ?? []
    }
}

/// Unstyled base variant (spec §8.3).
public struct BaseSignIn<Content: View>: View {
    @EnvironmentObject private var state: ThunderState
    public let applicationId: String
    public let onComplete: (() -> Void)?
    public let onError: ((String) -> Void)?
    public let content: (SignInState) -> Content

    @StateObject private var signInState = SignInState(submit: { _, _, _ in })

    public init(
        applicationId: String,
        onComplete: (() -> Void)? = nil,
        onError: ((String) -> Void)? = nil,
        @ViewBuilder content: @escaping (SignInState) -> Content
    ) {
        self.applicationId = applicationId
        self.onComplete = onComplete
        self.onError = onError
        self.content = content
    }

    public var body: some View {
        content(signInState)
            .task { await initFlow() }
    }

    private func initFlow() async {
        signInState.isLoading = true
        defer { signInState.isLoading = false }
        do {
            let request = EmbeddedFlowRequestConfig(applicationId: applicationId, flowType: .authentication)
            let payload = EmbeddedSignInPayload(actionId: "__initiate__")
            let response = try await state.client.signIn(payload: payload, request: request)
            await handleResponse(response)
        } catch {
            signInState.error = error.localizedDescription
            onError?(error.localizedDescription)
        }
    }

    private func submit(actionId: String, inputs: [String: String], flowId: String?) async {
        signInState.isLoading = true
        defer { signInState.isLoading = false }
        do {
            let payload = EmbeddedSignInPayload(flowId: flowId, actionId: actionId, inputs: inputs)
            let request = EmbeddedFlowRequestConfig(applicationId: applicationId, flowType: .authentication)
            let response = try await state.client.signIn(payload: payload, request: request)
            await handleResponse(response)
        } catch {
            signInState.error = error.localizedDescription
            onError?(error.localizedDescription)
        }
    }

    private func handleResponse(_ response: EmbeddedFlowResponse) async {
        switch response.flowStatus {
        case .complete:
            await state.refresh()
            onComplete?()
        case .promptOnly:
            signInState.update(from: response)
        case .error:
            let msg = response.failureReason ?? "Sign-in failed"
            signInState.error = msg
            onError?(msg)
        }
    }
}
